File: 0943-sum-of-subarray-minimums/0943-sum-of-subarray-minimums.java
class Solution {
    public int sumSubarrayMins(int[] arr) {
        int n = arr.length;
        ArrayList <Integer> NSL = getNsl(arr,n);
        ArrayList <Integer> RSL = getRsl(arr,n);
        long sum = 0;
        int M = 1000000007;
        for(int i = 0;i<n;i++){
            long ls = NSL.get(i);
            long rs = RSL.get(i);
            long totalWays = ls*rs;
            long totalSum = totalWays*arr[i];
            sum = (sum + totalSum)%M;
        }
        return (int) sum;
    }

    public ArrayList <Integer> getNsl(int [] arr,int n){
    Stack <Integer> s = new Stack<>();
    ArrayList <Integer> result = new ArrayList <>();
    for(int i = 0;i<n;i++){
        while(s.size() != 0 && arr[i] <= arr[s.peek()]){
            s.pop();
        }
        int nsl = s.isEmpty() ? i + 1 : i - s.peek();
        s.push(i);
        result.add(nsl);
    }
    return result;

    }

    public ArrayList <Integer> getRsl(int [] arr,int n){
    Stack <Integer> s = new Stack<>();
    ArrayList <Integer> result = new ArrayList <>();
    for(int i = n-1;i>=0;i--){
        while(s.size() != 0 && arr[i] < arr[s.peek()]){
            s.pop();
        }
         int rsl = s.isEmpty() ? n - i : s.peek() - i;
        s.push(i);
        result.add(rsl);
    }
     Collections.reverse(result);
        return result;

    }
}

File: 0943-sum-of-subarray-minimums/0943-sum-of-subarray-minimums.js
/**
 * @param {number[]} arr
 * @return {number}
 */
// var sumSubarrayMins = function(arr) {
//     let NSL = getNSL(arr);
//     let NSR = getNSR(arr);

//     let sum =0;
//     let m = 1e9 + 7;

//     for(let i = 0;i<arr.length;i++){
//         let left = i - NSL[i];
//         let right = NSR[i] - i;
//         let totalWays = left*right;
//         let totalSum = totalWays*arr[i];
//         sum = (sum+totalSum)%m;
//     }
//     return sum;
// };

var sumSubarrayMins =  function (nums){
    let nsl = getNsl(nums);
    let nsr = getNsr(nums);
    let sum = 0;
    let m = 1e9 + 7;

    for(let i = 0;i<nums.length;i++){
        let left = i - nsl[i];
        let right = nsr[i] - i;
        let totalWays = left*right;
        let totalSum = totalWays*nums[i];
         sum = (sum+totalSum)%m;
    }
    return sum;
}
var getNsl = function(nums){
    let st = [];
    let res = [];
    let n = nums.length;
    for(let i = 0;i<n;i++){
        while(st.length !== 0 && nums[st[st.length - 1]]>=nums[i]){
            st.pop();
           
        }
         res[i] = st.length === 0 ? -1 : st[st.length - 1];
         st.push(i);

    }
    return res;

}

var getNsr = function(nums){
    let st = [];
    let res = [];
    let n = nums.length;
    for(let i = n-1;i>=0;i--){
        while(st.length !== 0 && nums[st[st.length - 1]]>nums[i]){
            st.pop();
           
        }
         res[i] = st.length === 0 ? n : st[st.length - 1];
         st.push(i);

    }
    return res;

}